extensions/mapstore/src/main/java/com/hazelcast/mapstore/GenericMapStore.java

```diff
@@ -0,0 +1,518 @@
+                        + ")")) {
+        }
+    }
+
+    private SqlRowMetadata loadMetadataFromMapping(String mapping) {
+        try (SqlResult result = sql.execute("SELECT * FROM \"" + mapping + "\" LIMIT 0")) {
+            SqlRowMetadata rowMetadata = result.getRowMetadata();
+            return rowMetadata;
+        }
+    }
+
+    private void readExistingMapping() {
+        try (SqlResult mappings = sql.execute("SHOW MAPPINGS")) {
+            for (SqlRow mapping : mappings) {
+                String name = mapping.getObject(MAPPING_NAME_COLUMN);
+                if (name.equals(this.mapping)) {
+                    SqlRowMetadata rowMetadata = loadMetadataFromMapping(name);
+                    validateColumns(rowMetadata);
+                    columnMetadata = rowMetadata.getColumns();
+                    queries = new Queries(name, properties.idColumn, columnMetadata);
+                    initFinished.countDown();
+                    return;
+                }
+            }
+        } catch (Exception e) {
+            initFailure = e;
+            initFinished.countDown();
+        }
+    }
+
+    private void validateColumns(SqlRowMetadata rowMetadata) {
+        Stream.concat(of(properties.idColumn), properties.columns.stream())
+              .distinct() // avoid duplicate id column if present in columns property
+              .forEach((columnName) -> {
+                  int column = rowMetadata.findColumn(columnName);
+                  if (column == COLUMN_NOT_FOUND) {
+                      throw new HazelcastException("Column '" + columnName + "' not found");
+                  }
+              });
+    }
+
+    @Override
+    public void destroy() {
+        if (isMaster()) {
+            dropMapping(mapping);
+        }
+    }
+
+    @Override
+    public GenericRecord load(K key) {
+        awaitInitFinished();
+
+        try (SqlResult queryResult = sql.execute(queries.load(), key)) {
+            Iterator<SqlRow> it = queryResult.iterator();
+            if (it.hasNext()) {
+                SqlRow row = it.next();
+                if (it.hasNext()) {
+                    throw new IllegalStateException("multiple matching rows for a key " + key);
+                }
+                return convertRowToGenericRecord(key, row);
+            } else {
+                return null;
+            }
+        }
+    }
+
+    @Nonnull
+    private GenericRecord convertRowToGenericRecord(K key, SqlRow row) {
+        GenericRecordBuilder builder = GenericRecordBuilder.compact(properties.compactTypeName);
+
+        SqlRowMetadata metadata = row.getMetadata();
+        for (int i = 0; i < metadata.getColumnCount(); i++) {
+            SqlColumnMetadata column = metadata.getColumn(i);
+
+            if (column.getName().equals(properties.idColumn) && !properties.idColumnInColumns) {
+                continue;
+            }
+
+            switch (column.getType()) {
+                case VARCHAR:
```

Comment:
do we really want to support only fields with VARCHAR and INTEGER? Or is it for demo and other types will be implemented later?

Reply:
This was a test if the reviewers are careful enough ;-)

00e5aa7a